File: Libraries/graphics/transformation/transformation.cpp
#include <cstring>
#include "transformation.hpp"

void gfx::_Transformation::stretch(float x, float y) {
    float applied_matrix[3][3] = {
        {x,   0.f, 0.f},
        {0.f, y,   0.f},
        {0.f, 0.f, 1.f},
    };
    *this = *this * _Transformation(applied_matrix);
}

void gfx::_Transformation::translate(float x, float y) {
    float applied_matrix[3][3] = {
        {1.f, 0.f, 0.f},
        {0.f, 1.f, 0.f},
        {x,   y,   1.f},
    };
    *this = *this * _Transformation(applied_matrix);
}

const float* gfx::_Transformation::getArray() const {
    return &matrix[0][0];
}

gfx::_Transformation gfx::_Transformation::operator*(const _Transformation& a) {
    _Transformation result;
    for(int x = 0; x < 3; x++)
        for(int y = 0; y < 3; y++) {
            float sum = 0;
            for(int i = 0; i < 3; i++)
                sum += matrix[i][y] * a.matrix[x][i];
            result.matrix[x][y] = sum;
        }
    return result;
}

gfx::_Transformation::_Transformation(float matrix_[3][3]) {
    for(int i = 0; i < 3; i++)
        for(int j = 0; j < 3; j++)
            matrix[i][j] = matrix_[i][j];
}

File: libraries/graphics/transformation.rs
use crate::libraries::graphics as gfx;

/// Transformation is a 3x3 matrix used to move and scale drawn objects.
#[derive(Clone)]
pub struct Transformation {
    pub matrix: [f32; 9],
}

impl Transformation {
    /// A new Transformation always contains identity matrix.
    pub const fn new() -> Self {
        Self {
            matrix: [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0],
        }
    }

    /// Moves the object around.
    pub(super) fn translate(&mut self, pos: gfx::FloatPos) {
        self.matrix[6] += self.matrix[0] * pos.0 + self.matrix[3] * pos.1;
        self.matrix[7] += self.matrix[1] * pos.0 + self.matrix[4] * pos.1;
        self.matrix[8] += self.matrix[2] * pos.0 + self.matrix[5] * pos.1;
    }

    /// Scales the object.
    pub(super) fn stretch(&mut self, scale: (f32, f32)) {
        self.matrix[0] *= scale.0;
        self.matrix[1] *= scale.0;
        self.matrix[2] *= scale.0;
        self.matrix[3] *= scale.1;
        self.matrix[4] *= scale.1;
        self.matrix[5] *= scale.1;
    }
}
